packages/@lwc/engine/src/framework/upgrade.ts

```diff
@@ -102,50 +89,32 @@ export function createElement(sel: string, options: CreateElementOptions): HTMLE
         );
     }
 
-    let Ctor = options.is;
+    const Ctor = options.is;
     if (!isFunction(Ctor)) {
         throw new TypeError(
             `"createElement" function expects a "is" option with a valid component constructor.`
         );
     }
 
-    const mode = options.mode !== 'closed' ? 'open' : 'closed';
-
-    // Create element with correct tagName
     const element = document.createElement(sel);
+
+    // There is a possibility that a custom element is registered under tagName, in which case, the
+    // initialization is already carry on, and there is nothing else to do here.
     if (!isUndefined(getAssociatedVMIfPresent(element))) {
-        // There is a possibility that a custom element is registered under tagName,
-        // in which case, the initialization is already carry on, and there is nothing else
-        // to do here.
         return element;
     }
 
-    if (isCircularModuleDependency(Ctor)) {
-        Ctor = resolveCircularModuleDependency(Ctor);
-    }
-
     const def = getComponentDef(Ctor);
     setElementProto(element, def);
 
-    if (process.env.NODE_ENV !== 'production') {
-        patchCustomElementWithRestrictions(element, EmptyObject);
-    }
-    // In case the element is not initialized already, we need to carry on the manual creation
-    createVM(element, Ctor, { mode, isRoot: true, owner: null });
-    // Handle insertion and removal from the DOM manually
-    setHiddenField(element, ConnectingSlot, () => {
-        const vm = getAssociatedVM(element);
-        startGlobalMeasure(GlobalMeasurementPhase.HYDRATE, vm);
-        if (vm.state === VMState.connected) {
-            // usually means moving the element from one place to another, which is observable via life-cycle hooks
-            removeRootVM(vm);
-        }
-        appendRootVM(vm);
-        endGlobalMeasure(GlobalMeasurementPhase.HYDRATE, vm);
-    });
-    setHiddenField(element, DisconnectingSlot, () => {
-        const vm = getAssociatedVM(element);
-        removeRootVM(vm);
+    const vm = createVM(element, def.ctor, {
+        mode: options.mode !== 'closed' ? 'open' : 'closed',
+        isRoot: true,
+        owner: null,
     });
+
+    setHiddenField(element, ConnectingSlot, () => appendRootVM(vm));
```

Comment:
The VM object has a strong reference to the element. 
Storing the relationship between element and the arrow function leaves a reference to the vm due to closure. I am wondering if this leads to a circular reference and as a result a memory leak. 🤔 
_`setHiddenField` AFAIK uses a WeakMap, but I am unclear if the values of a WeakMap are GC eligible._

Reply:
I think @ravijayaramappa is right, let's use the `this` value inside that method to extract the `vm` our of it, just to make sure that they don't call that arbitrarily. the side effect of that is that you can copy that method to any element, and it will attempt to append the root vm. Maybe that's not a big deal. Also, we can cache this method instead of creating one every time.